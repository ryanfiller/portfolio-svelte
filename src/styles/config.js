// http://ecoiron.blogspot.com/2007/01/emergy-c-low-wattage-palette.html

const emergyc = {
  // #822007 (rusty red)
  // #000000 (black)
  // #b2bbc0 (blue grey)
  // #19472a (forest green)
  // #3d414c (cobalt)
  // #ffffff (white)
}

export const colors = {
  black: '#080025',
  white: '#fefdf2',
  purpleLight: '#a176b6',
  purpleDark: '#5651a7',
  blueLight: '#4f81c0',
  blueDark: '#155f91',
  orangeLight: '#f16a1f',
  orangeDark: '#ac284f',
  grayLight: '#5e828f',
  grayDark: '#4e6773',
}

export const themes = {
  light: {
    text: 'black',
    background: 'white',
    primary: 'purpleDark',
    highlight: 'blueDark',
    active: 'orangeLight',
    disabled: 'grayLight',
  },
  dark: {
    text: 'white',
    background: 'black',
    primary: 'purpleDark',
    highlight: 'orangeDark',
    active: 'blueLight',
    disabled: 'grayDark',
  }

  // light: {
  //   text: 'black',
  //   background: 'white',
  //   primary: 'purpleDark',
  //   highlight: 'blueDark',
  //   active: 'orangeLight',
  //   disabled: 'grayLight',
  // },
  // dark: {
  //   text: 'white',
  //   background: 'black',
  //   primary: 'purpleDark',
  //   highlight: 'orangeDark',
  //   active: 'blueLight',
  //   disabled: 'grayDark',
  // }
}

export const fonts = {
  plex: {
    name: 'IBM Plex',
    url: 'https://www.ibm.com/plex/',
      options: {
      wght: [100, 700],
      wdth: [85, 100]
    }
  },

  'science-gothic': {
    name: 'Science Gothic',
    url: 'https://github.com/tphinney/science-gothic',
    options: {
      wght: [100, 900],
      wdth: [50, 200],
      YOPQ: [18, 122]
    }
  },

  fira: {
    name: 'Fira Code',
    url: 'https://github.com/tonsky/FiraCode',
      options: {
      wght: [300, 700]
    }
  }
}
