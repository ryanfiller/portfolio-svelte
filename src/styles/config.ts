export type Fonts = {
  [key: string]: Font
}

export type Font = {
  url: string,
  options: {
    [key: string]: [number, number]
  }
}

export const fonts: Fonts = {  
  // headers fonts

  'Science Gothic': {
    url: 'https://github.com/tphinney/science-gothic',
    options: {
      wght: [100, 900],
      wdth: [50, 200],
      YOPQ: [18, 122]
    }
  },

  'Fraunces': {
    url: 'https://fraunces.undercase.xyz/',
    options: {
      SOFT: [0, 100],
      wght: [100, 900],
      WONK: [0, 1],
      opsz: [9, 144]
    }
  },
  
  // body fonts

  'Inter': {
    url: 'https://rsms.me/inter/',
    options: {
      wght: [100, 900],
      slnt: [-10, 0]
    }
  },

  'IBM Plex': {
    url: 'https://www.ibm.com/plex/',
    options: {
      wght: [100, 700],
      wdth: [85, 100]
    }
  },

  // code fonts

  'Recursive': {
    url: 'https://www.recursive.design/',
    options: {
      MONO: [0, 1],
      CASL: [0, 1],
      wght: [300, 1000],
      slnt: [-15, 0],
      CRSV: [0, 1]
    }
  },

  'Fira Code': {
    url: 'https://github.com/tonsky/FiraCode',
    options: {
      wght: [300, 700]
    }
  }
}
