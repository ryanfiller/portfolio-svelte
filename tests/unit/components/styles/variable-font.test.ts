import { describe, it, expect, vi, beforeEach, afterEach } from 'vitest';
import { render, fireEvent, waitFor, within, isInaccessible } from '@testing-library/svelte';
import type { RenderResult } from '@testing-library/svelte';

import { mockMatchMedia } from '../../test-helpers';

import { fonts } from '$styles/config';
import VariableFont from '$components/styles/variable-font.svelte';

// TODO - e2e test this component

describe('<VariableFont />', () => {
	let component: RenderResult<VariableFont>;
	let textArea: HTMLTextAreaElement,
		variationOptions: HTMLInputElement[],
		italic: HTMLInputElement,
		textTransform: HTMLSelectElement,
		css: HTMLElement;
	// @ts-ignore
	let performance;

	// TODO - should this props function be abstracted to a helper?
	function props(overrides = {}) {
		return {
			fontName: Object.keys(fonts)[0],
			...overrides
		};
	}

	function createComponent() {
		component = render(VariableFont, props());
		textArea = component.getByRole('textbox') as HTMLTextAreaElement;
		const inputs = component.getByRole('group');
		variationOptions = within(inputs).getAllByRole('slider');
		italic = within(inputs).getByRole('checkbox');
		textTransform = within(inputs).getByRole('combobox');
		css = component.getByRole('code');
	}

	beforeEach(() => {
		// lots of complicated mocking for this component...

		mockMatchMedia('(prefers-reduced-data: reduce)', false);

		vi.mock('$styles/fonts.css?inline', () => ({
			default: `
        @font-face {
          font-family: '${Object.keys(fonts)[0]}';
          src: url('/fonts/whatever.ttf');
        }

				:root {
					--test-font: '${Object.keys(fonts)[0]}';
				}
      `
		}));

		vi.mock('$app/environment', () => ({
			browser: true
		}));

		performance = global.performance;

		global.performance = {
			getEntriesByType: () => [
				// @ts-ignore
				{ name: Object.keys(fonts)[0].replace(' ', '-'), decodedBodySize: 1024 }
			]
		};

		createComponent();
	});

	afterEach(() => {
		vi.resetAllMocks();

		// @ts-ignore
		global.performance = performance;
	});

	describe('rendering', () => {
		it('renders consistently', () => {
			expect(component.container).toMatchSnapshot();
		});

		it('ia accessible', () => {
			expect(isInaccessible(component.container)).toBe(false);
		});

		it('renders the font name and size', () => {
			const font = Object.keys(fonts)[0];
			expect(component.getByRole('banner').textContent).toMatch(`--test-font`);
			expect(component.getByRole('banner').textContent).toMatch(`${font} (1.00KB)`)
		});

		it('renders the font options', async () => {
			expect(variationOptions.length).toEqual(
				Object.entries(Object.values(fonts)[0].options).length
			);
			expect(italic).toBeTruthy();
			expect(textTransform).toBeTruthy();
		});
	});

	describe('interacting', () => {
		const property = Object.keys(Object.values(fonts)[0].options)[0];
		const min = Object.values(Object.values(fonts)[0].options)[0][0];

		beforeEach(() => {
			fireEvent.input(textArea, {
				target: { value: 'the quick brown fox jumped over the lazy dog' }
			});
			fireEvent.change(variationOptions[0], { target: { value: min } });
			fireEvent.change(italic, { target: { checked: true } });
			fireEvent.change(textTransform, { target: { value: 'lowercase' } });
		});

		it('changes the example text', () => {
			expect(textArea.value).toMatch('the quick brown fox jumped over the lazy dog');
		});

		it('applies the styles', () => {
			const inlineStyles = (textArea.parentElement as HTMLElement)?.getAttribute('style');
			expect(inlineStyles).toMatch(`'${property}' ${min}`);
			expect(inlineStyles).toMatch(`font-style: italic`);
			expect(inlineStyles).toMatch(`text-transform: lowercase`);
		});

		it('creates the css', () => {
			const font = Object.keys(fonts)[0];
			expect(css.textContent).toMatch(`font-family: "${font}";`);
			expect(css.textContent).toMatch(`"${property}" ${min}`);
			expect(css.textContent).toMatch(`font-style: italic;`);
			expect(css.textContent).toMatch(`text-transform: lowercase;`);
		});
	});

	describe('prefers-reduced-data', () => {
		beforeEach(() => {
			mockMatchMedia('(prefers-reduced-data: reduce)', true);
			// unmount the original component
			component.unmount();
			// recreate it with reduced data
			createComponent();
		});

		it('filters the font options correctly', () => {
			Object.keys(Object.values(fonts)[0].options).forEach((option) => {
				if (option === 'wght' || option === 'wdth') {
					expect(component.getByLabelText(option)).toBeTruthy();
				} else {
					expect(() => component.getByLabelText(option)).toThrow();
				}
			});
		});

		it('applies the styles', () => {
			const inlineStyles = (textArea.parentElement as HTMLElement)?.getAttribute('style');
			expect(inlineStyles).toMatch(`--reduced-data-font: var(--test-font);`);
		});

		it('creates the css', () => {
			expect(css.textContent).toMatch(`font-family: var(--reduced-data-font);`);
		});
	});
});
