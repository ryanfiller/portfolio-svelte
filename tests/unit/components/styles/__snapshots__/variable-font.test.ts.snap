// Vitest Snapshot v1

exports[`<VariableFont /> > rendering > renders consistently 1`] = `
<body>
  <div>
    <section
      class="variable-font s-mErrwtoW59yo"
    >
      <header
        class="s-mErrwtoW59yo"
      >
        <code
          class="s-mErrwtoW59yo"
        >
          --test-font
        </code>
         
        <span
          class="no-reduce-data s-mErrwtoW59yo"
        >
          <a
            class="s-mErrwtoW59yo"
            href="https://sciencegothic.com/"
            rel="noopener noreferrer"
            target="_blank"
          >
            Science Gothic
          </a>
           
          <span
            class="needs-js s-mErrwtoW59yo"
          >
            (...)
          </span>
        </span>
      </header>
       
      <fieldset
        class="needs-js s-mErrwtoW59yo"
      >
        <legend
          class="s-mErrwtoW59yo"
        >
          Options:
        </legend>
         
        <div
          class="slider s-mErrwtoW59yo"
        >
          <label
            class="s-mErrwtoW59yo"
            for="science-gothic-wght-slider"
          >
            wght
          </label>
           
          <input
            class="s-mErrwtoW59yo"
            id="science-gothic-wght-slider"
            max="900"
            min="100"
            name="wght"
            step="1"
            type="range"
          />
        </div>
        <div
          class="slider s-mErrwtoW59yo"
        >
          <label
            class="s-mErrwtoW59yo"
            for="science-gothic-wdth-slider"
          >
            wdth
          </label>
           
          <input
            class="s-mErrwtoW59yo"
            id="science-gothic-wdth-slider"
            max="200"
            min="50"
            name="wdth"
            step="1"
            type="range"
          />
        </div>
        <div
          class="slider s-mErrwtoW59yo"
        >
          <label
            class="s-mErrwtoW59yo"
            for="science-gothic-YOPQ-slider"
          >
            YOPQ
          </label>
           
          <input
            class="s-mErrwtoW59yo"
            id="science-gothic-YOPQ-slider"
            max="122"
            min="18"
            name="YOPQ"
            step="1"
            type="range"
          />
        </div>
         
        <div
          class="checkbox s-mErrwtoW59yo"
        >
          <label
            class="s-mErrwtoW59yo"
            for="science-gothic-italic"
          >
            italic
          </label>
           
          <input
            class="s-mErrwtoW59yo"
            id="science-gothic-italic"
            type="checkbox"
          />
        </div>
         
        <div
          class="select s-mErrwtoW59yo"
        >
          <label
            class="s-mErrwtoW59yo"
            for="science-gothic-capitalization"
          >
            text-transform
          </label>
           
          <select
            class="s-mErrwtoW59yo"
            id="science-gothic-capitalization"
          >
            <option
              class="s-mErrwtoW59yo"
              value="none"
            >
              none
            </option>
            <option
              class="s-mErrwtoW59yo"
              value="capitalize"
            >
              capitalize
            </option>
            <option
              class="s-mErrwtoW59yo"
              value="uppercase"
            >
              uppercase
            </option>
            <option
              class="s-mErrwtoW59yo"
              value="lowercase"
            >
              lowercase
            </option>
          </select>
        </div>
      </fieldset>
       
      <pre
        class="code needs-js s-mErrwtoW59yo"
        role="code"
      >
        		
        <code
          class="s-mErrwtoW59yo"
        >
          
			font-family: 
          "Science Gothic"
          ;
			font-variation-settings: 
          
  "wght" 500,
  "wdth" 125,
  "YOPQ" 70
          ;
			
          
			
          
		
        </code>
        
	
      </pre>
       
      <div
        class="example s-mErrwtoW59yo"
      >
        <label
          class="screen-reader s-mErrwtoW59yo"
          for="science-gothic-example-text"
        >
          text example
        </label>
         
        <div
          class="s-mErrwtoW59yo"
          style="font-family: Science Gothic; font-variation-settings: 'wght' 500, 'wdth' 125, 'YOPQ' 70; text-transform: none; font-style: unset; /* lol, wut */ /* stylelint disable custom-property-empty-line-before */ --reduced-data-font: var(--test-font);"
        >
          <textarea
            class="example s-mErrwtoW59yo"
            id="science-gothic-example-text"
            wrap="hard"
          />
           
          <p
            class="s-mErrwtoW59yo"
          >
            a b c d e f g h i j k l m n o p q r s t u v w x y z A B C D E F G H I J K L M N O P Q R S T U V W X Y Z
             
            <br
              class="s-mErrwtoW59yo"
            />
             
            0 1 2 3 4 5 6 7 8 9
             
            <br
              class="s-mErrwtoW59yo"
            />
             
            ! " # $ % & ' ( ) * + , - . / : ; &lt; = &gt; ? @ { | } ~ [ \\ ] ^ _ \`
             
            <br
              class="s-mErrwtoW59yo"
            />
             
            && || =&gt; == != &lt;= &gt;= !! ?.
             
            <br
              class="s-mErrwtoW59yo"
            />
          </p>
        </div>
      </div>
    </section>
    <!--&lt;Variable-font&gt;-->
  </div>
</body>
`;
